File: MarvinPlugin/Source/MarvinPlugin/XcodeManager.m
//
//  XcodeManager.m
//  MarvinPlugin
//
//  Created by Christoffer Winterkvist on 17/09/14.
//  Based on implementations by Benoît Bourdon
//  Copyright (c) 2014 zenangst. All rights reserved.
//

#import "XcodeManager.h"

@interface XcodeManager ()

@property (nonatomic, strong) NSTextView *textView;

@end

@implementation XcodeManager

#pragma mark - Getters

- (id)currentEditor
{
    NSWindowController *currentWindowController = [[NSApp keyWindow] windowController];

    if ([currentWindowController isKindOfClass:NSClassFromString(@"IDEWorkspaceWindowController")]) {
        IDEWorkspaceWindowController *workspaceController = (IDEWorkspaceWindowController *)currentWindowController;
        IDEEditorArea *editorArea = [workspaceController editorArea];
        IDEEditorContext *editorContext = [editorArea lastActiveEditorContext];
        return [editorContext editor];
    }

    return nil;
}

- (NSTextView *)textView
{
    if ([[self currentEditor] isKindOfClass:NSClassFromString(@"IDESourceCodeEditor")]) {
        IDESourceCodeEditor *editor = [self currentEditor];
        return editor.textView;
    }

    if ([[self currentEditor] isKindOfClass:NSClassFromString(@"IDESourceCodeComparisonEditor")]) {
        IDESourceCodeComparisonEditor *editor = [self currentEditor];
        return editor.keyTextView;
    }

    return nil;
}

- (IDESourceCodeDocument *)currentSourceCodeDocument
{
    if ([[self currentEditor] isKindOfClass:NSClassFromString(@"IDESourceCodeEditor")]) {
        IDESourceCodeEditor *editor = [self currentEditor];
        return editor.sourceCodeDocument;
    }

    if ([[self currentEditor] isKindOfClass:NSClassFromString(@"IDESourceCodeComparisonEditor")]) {
        IDESourceCodeComparisonEditor *editor = [self currentEditor];
        if ([[editor primaryDocument] isKindOfClass:NSClassFromString(@"IDESourceCodeDocument")]) {
            IDESourceCodeDocument *document = (IDESourceCodeDocument *)editor.primaryDocument;
            return document;
        }
    }

    return nil;
}

- (IDEEditorDocument *)currentDocument
{
    NSWindowController *currentWindowController = [[NSApp keyWindow] windowController];

    if ([currentWindowController isKindOfClass:NSClassFromString(@"IDEWorkspaceWindowController")]) {
        IDEWorkspaceWindowController *workspaceController = (IDEWorkspaceWindowController *)currentWindowController;
        IDEEditorArea *editorArea = [workspaceController editorArea];
        return editorArea.primaryEditorDocument;
    }

    return nil;
}

- (void)save
{
    if ([[self currentSourceCodeDocument] isEqualTo:[self currentDocument]]) {
        [[self currentDocument] saveDocument:nil];
    } else {
        [[self currentSourceCodeDocument] saveDocument:nil];
    }
}

- (void)needsDisplay
{
    [self.textView setNeedsDisplay:YES];
}

- (NSString *)contents
{
    return [self.textView string];
}

- (NSUInteger)documentLength
{
    return [[self contents] length];
}

- (NSRange)selectedRange
{
    return self.textView.selectedRange;
}

- (NSRange)currentWordRange
{
    NSCharacterSet *validSet = [NSCharacterSet characterSetWithCharactersInString:kMarvinValidSetWordString];
    NSCharacterSet *spaceSet = [NSCharacterSet characterSetWithCharactersInString:kMarvinSpaceSet];
    NSRange selectedRange = [self selectedRange];

    BOOL isOutOfBounds = (selectedRange.location >= self.contents.length);
    if (!isOutOfBounds) {
        char character;
        if ([self hasSelection]) {
            character = [[self contents] characterAtIndex:selectedRange.location+selectedRange.length];
        } else {
            character = [[self contents] characterAtIndex:selectedRange.location];
        }

        if (![validSet characterIsMember:character]) {
            selectedRange = (NSRange) { .location = selectedRange.location + selectedRange.length };
        }

        NSScanner *scanner = [NSScanner scannerWithString:[self contents]];
        [scanner setScanLocation:selectedRange.location];

        NSUInteger length = selectedRange.location;

        while (!scanner.isAtEnd) {
            if ([scanner scanCharactersFromSet:validSet
                                    intoString:nil]) {
                length = [scanner scanLocation];
                break;
            }
            [scanner setScanLocation:[scanner scanLocation] + 1];
        }

        NSUInteger location = ([[self contents] rangeOfCharacterFromSet:spaceSet
                                                                options:NSBackwardsSearch
                                                                  range:NSMakeRange(0,length)].location +1);

        if (length-location > self.documentLength) {
            length = 0;
        }

        return NSMakeRange(location,length-location);
    } else {
        return selectedRange;
    }
}

- (NSRange)previousWordRange
{
    NSRange selectedRange = [self selectedRange];
    NSCharacterSet *validSet = [NSCharacterSet characterSetWithCharactersInString:kMarvinValidSetWordString];
    NSUInteger location = ([[self contents] rangeOfCharacterFromSet:validSet
                                                            options:NSBackwardsSearch
                                                              range:NSMakeRange(0,selectedRange.location)].location);

    return NSMakeRange(location,0);
}

- (NSRange)lineContentsRange
{
    NSRange selectedRange = [self selectedRange];

    NSCharacterSet *newlineSet = [NSCharacterSet characterSetWithCharactersInString:@"\n"];
    NSUInteger startOfLine = ([[self contents] rangeOfCharacterFromSet:newlineSet
                                                               options:NSBackwardsSearch
                                                                 range:NSMakeRange(0,selectedRange.location)].location);

    NSCharacterSet *validSet = [NSCharacterSet characterSetWithCharactersInString:kMarvinValidLineRange];

    if (startOfLine == NSNotFound) startOfLine = 0;

    NSUInteger location = ([[self contents] rangeOfCharacterFromSet:validSet
                                                            options:NSCaseInsensitiveSearch
                                                              range:NSMakeRange(startOfLine,[self documentLength]-startOfLine)].location);

    NSUInteger length = ([[self contents] rangeOfCharacterFromSet:newlineSet
                                                          options:NSCaseInsensitiveSearch
                                                            range:NSMakeRange(selectedRange.location+selectedRange.length,[self contents].length-(selectedRange.location+selectedRange.length))].location);

    if (length-location < [self documentLength]) {
        return NSMakeRange(location, length-location);
    } else {
        return NSMakeRange(selectedRange.location, 0);
    }
}

- (NSRange)lineRange
{
    NSRange selectedRange = [self selectedRange];
    NSCharacterSet *newlineSet = [NSCharacterSet characterSetWithCharactersInString:@"\n"];
    NSUInteger location = ([[self contents] rangeOfCharacterFromSet:newlineSet
                                                            options:NSBackwardsSearch
                                                              range:NSMakeRange(0,selectedRange.location)].location);

    NSUInteger length = ([[self contents] rangeOfCharacterFromSet:newlineSet
                                                          options:NSCaseInsensitiveSearch
                                                            range:NSMakeRange(selectedRange.location+selectedRange.length,[self contents].length-(selectedRange.location+selectedRange.length))].location);

    location = (location == NSNotFound) ? 0 : location + 1;
    length   = (location == 0) ? length+1   : (length+1) - location;

    return NSMakeRange(location, length);
}

- (NSString *)contentsOfRange:(NSRange)range
{
    return [[self contents] substringWithRange:range];
}

- (NSRange)joinRange
{
    NSRange lineRange = [self lineRange];
    NSRange joinRange = (NSRange) { .location = lineRange.location + lineRange.length - 1 };

    NSCharacterSet *validSet = [NSCharacterSet characterSetWithCharactersInString:kMarvinValidLineRange];

    NSUInteger length = ([[self contents] rangeOfCharacterFromSet:validSet
                                                          options:NSCaseInsensitiveSearch
                                                            range:NSMakeRange(joinRange.location,[self contents].length-joinRange.location)].location);

    return NSMakeRange(joinRange.location, length - joinRange.location);
}

- (NSString *)selectedText
{
    NSString *text = [[self.textView string] substringWithRange:self.textView.selectedRange];
    return text;
}

- (BOOL)hasSelection
{
    return (self.textView.selectedRange.length) ? YES : NO;
}

- (BOOL)emptySelection
{
    return (![self hasSelection]);
}

- (NSLayoutManager *)layoutManager
{
    return self.textView.layoutManager;
}

#pragma mark - Setters

- (void)insertText:(NSString *)string
{
    [self.textView insertText:string];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [[NSNotificationCenter defaultCenter] postNotificationName:@"Add change mark"
                                                            object:string];
    });
}

- (void)setSelectedRange:(NSRange)range
{
    if (range.location != NSNotFound) {
        if ((range.location + range.length) > self.contents.length) {
            range.length = self.contents.length - range.location;
        }

        self.textView.selectedRange = range;
    }
}

- (void)replaceCharactersInRange:(NSRange)range
                      withString:(NSString *)string
{
    if ((range.location + range.length) > self.contents.length) {
        range.length = self.contents.length - range.location;
    }

    IDESourceCodeDocument *document = [self currentSourceCodeDocument];
    DVTSourceTextStorage *textStorage = [document textStorage];

    [textStorage replaceCharactersInRange:range
                               withString:string
                          withUndoManager:[document undoManager]];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [[NSNotificationCenter defaultCenter] postNotificationName:@"Add change mark"
                                                            object:string];
    });
}

@end

File: MarvinPlugIn/Source/MarvinPlugIn/XcodeManager.swift
import Foundation

class XcodeManager: NSObject {

  var textView: NSTextView? {
    get {
      if let currentEditor = LegacyXcodeManager().currentEditor(),
        let className = NSClassFromString("IDESourceCodeEditor")
        , (currentEditor as AnyObject).isKind(of: className) {
          return (currentEditor as AnyObject).textView
      }

      if let currentEditor = LegacyXcodeManager().currentEditor(),
        let className = NSClassFromString("IDESourceCodeComparisonEditor")
        , (currentEditor as AnyObject).isKind(of: className) {
          return (currentEditor as AnyObject).keyTextView
      }

      return nil
    }
  }

  var selectedRange: NSRange {
    set(value) {
      guard value.location != NSNotFound else { return }

      if value.location + value.length > self.contents().characters.count {
        var value = value
        value.length = self.contents().characters.count - value.location
      }

      textView?.selectedRange = value
    }
    get {
      return textView?.selectedRange ?? NSRange(location: 0, length: 0)
    }
  }

  func save() {
    LegacyXcodeManager().save()
  }
  func needsDisplay() {
    textView?.needsDisplay = true
  }

  func contents() -> String {
    return textView?.string ?? ""
  }

  func documentLength() -> Int {
    return (textView?.string?.characters.count ?? 0) - 1
  }

  func currentWordRange() -> NSRange {
    let validSet = CharacterSet(charactersIn: "0123456789ABCDEFGHIJKOLMNOPQRSTUVWXYZÅÄÆÖØabcdefghijkolmnopqrstuvwxyzåäæöø_")
    let spaceSet = CharacterSet(charactersIn: "#-<>/(){}[],;:. \n`*\"' ")
    var selectedRange = self.selectedRange

    guard selectedRange.location + selectedRange.length < contents().characters.count else { return selectedRange }

    var character: Character
    if self.hasSelection() {
      character = self.contents()[self.contents().characters.index(self.contents().startIndex, offsetBy: selectedRange.location+selectedRange.length)]
    } else {
      character = self.contents()[self.contents().characters.index(self.contents().startIndex, offsetBy: selectedRange.location)]
    }

    if !isChar(character, inSet:validSet) {
      selectedRange.location = selectedRange.location + selectedRange.length
    }

    let scanner = Scanner(string: self.contents())
    scanner.scanLocation = selectedRange.location

    var length = selectedRange.location

    while !scanner.isAtEnd {
      if scanner.scanCharacters(from: validSet, into: nil) {
        length = scanner.scanLocation
        break
      }

      scanner.scanLocation = scanner.scanLocation + 1
    }

    let whitespaceRange = (self.contents() as NSString).rangeOfCharacter(from: spaceSet,
      options: .backwards,
      range: NSRange(location: 0, length: length))

    let location = whitespaceRange.location != NSNotFound ? whitespaceRange.location + 1 : 0

    if length - location > self.documentLength() {
      length = 0
    }

    var range = NSRange(location: 0, length: 0)
    if location >= 0 {
      range = NSRange(location: location, length: length - location)
      return range
    } else if location == 0 && range.location != selectedRange.location && range.length != selectedRange.length {
      scanner.scanLocation = 0
      while !scanner.isAtEnd {
        if scanner.scanCharacters(from: validSet, into: nil) {
          length = scanner.scanLocation
          break
        }
        scanner.scanLocation = scanner.scanLocation + 1

        range.location = location
        range.length = length - location
      }

      if range.location == NSNotFound { range.location = 0 }

      if range.length > self.contents().characters.count {
        range.length = self.contents().characters.count
      }

      return range
    }

    return selectedRange
  }

  func previousWordRange() -> NSRange {
    let selectedRange = self.selectedRange
    let validSet = CharacterSet(charactersIn: "0123456789ABCDEFGHIJKOLMNOPQRSTUVWXYZÅÄÆÖØabcdefghijkolmnopqrstuvwxyzåäæöø_")
    var location = (self.contents() as NSString).rangeOfCharacter(from: validSet, options: .backwards, range: NSMakeRange(0,selectedRange.location)).location

    if location == NSNotFound {
      location = 0
    }

    return NSRange(location: location, length: 0)
  }

  func lineContentsRange() -> NSRange {
    let lineRange = self.lineRange()
    let currentLine = (self.contents() as NSString).substring(with: lineRange)
    let trimmedString = currentLine.trimmingCharacters(in: CharacterSet.whitespaces)
    let spacing = currentLine.replacingOccurrences(of: trimmedString, with: "")

    return NSRange(location: lineRange.location + spacing.characters.count, length: lineRange.length - spacing.characters.count - 1)
  }

  func lineRange() -> NSRange {
    var selectedRange = self.selectedRange

    if selectedRange.location == self.contents().characters.count {
      selectedRange.location -= 1
    }

    let newLineSet = CharacterSet(charactersIn: "\n")

    var location = (self.contents() as NSString).rangeOfCharacter(from: newLineSet, options: .backwards, range: NSRange(location: 0, length: selectedRange.location)).location
    var length = (self.contents() as NSString)
      .rangeOfCharacter(from: newLineSet,
        options: .caseInsensitive,
        range: NSRange(location: selectedRange.location + selectedRange.length,
          length: self.contents().characters.count - (selectedRange.location + selectedRange.length)))
      .location

    if length == NSNotFound {
      length = self.contents().characters.count - location
      return NSRange(location: location, length: length)
    }

    location = location == NSNotFound ? 0 : location + 1
    length = location == 0 ? length + 1 : length + 1 - location

    if length > self.contents().characters.count {
      length = self.contents().characters.count - location
    }

    return NSRange(location: location, length: length)
  }

  func contentsOfRange(_ range: NSRange) -> String {
    guard let textView = self.textView, let contents = textView.string else { return "" }
    return (contents as NSString).substring(with: range)
  }

  func joinRange() -> NSRange {
    let lineRange = self.lineRange()
    let joinRange = NSRange(location: lineRange.location + lineRange.length - 1, length: 0)
    let validSet = CharacterSet(charactersIn: "0123456789ABCDEFGHIJKOLMNOPQRSTUVWXYZÅÄÆÖØabcdefghijkolmnopqrstuvwxyzåäæöø_{}().$[]")
    let length = (self.contents() as NSString).rangeOfCharacter(from: validSet, options: .caseInsensitive, range: NSRange(location: joinRange.location, length: self.contents().characters.count - joinRange.location)).location

    return NSRange(location: joinRange.location, length: length - joinRange.location)
  }

  func selectedText() -> String {
    guard let textView = self.textView else { return "" }
    return contentsOfRange(textView.selectedRange)
  }

  func hasSelection() -> Bool {
    return self.textView?.selectedRange.length ?? 0 > 0
  }

  func emptySelection() -> Bool {
    return self.hasSelection() == false
  }

  func layoutManager() -> NSLayoutManager? {
    return self.textView?.layoutManager
  }

  func insertText(_ string: String) {
    self.textView?.insertText(string)

    let delayTime = DispatchTime.now() + 0.025
    DispatchQueue.main.asyncAfter(deadline: delayTime) {
      NotificationCenter.default.post(name: Notification.Name(rawValue: "Add change mark"), object: string)
    }
  }

  func replaceCharactersInRange(_ range: NSRange, withString string: String) {
    if range.location + range.length > self.contents().characters.count {
      var range = range
      range.length = self.contents().characters.count - range.location
    }

    let document = LegacyXcodeManager().currentSourceCodeDocument()
    let textStorage = document?.textStorage()

    textStorage?.replaceCharacters(in: range, with: string, withUndoManager: document?.undoManager)

    let delayTime = DispatchTime.now() + 0.025
    DispatchQueue.main.asyncAfter(deadline: delayTime) {
      NotificationCenter.default.post(name: Notification.Name(rawValue: "Add change mark"), object: string)
    }
  }

  fileprivate func isChar(_ char: Character, inSet set: CharacterSet) -> Bool {
    var found = false
    for ch in String(char).utf16 {
      if set.contains(UnicodeScalar(ch)!) { found = true; break }
    }
    return found
  }
}
